engines/mads/mps_installer.cpp

```diff
@@ -0,0 +1,160 @@
+/* ScummVM Tools
+ *
+ * ScummVM Tools is the legal property of its developers, whose names
+ * are too numerous to list here. Please refer to the COPYRIGHT
+ * file distributed with this source distribution.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "common/endian.h"
+#include "common/debug.h"
+#include "common/str.h"
+#include "common/file.h"
+#include "common/util.h"
+#include "common/memstream.h"
+#include "common/dcl.h"
+#include "mps_installer.h"
+
+namespace MADS {
+
+MpsInstaller* MpsInstaller::open(const Common::Path& baseName) {
+	Common::File indexFile;
+	Common::HashMap<Common::String, FileDescriptor, Common::IgnoreCase_Hash, Common::IgnoreCase_EqualTo> files;
+
+	if (!indexFile.open(baseName.append(".IDX")))
+		return nullptr;
+
+	uint32 filecnt = indexFile.readUint16LE();
+
+	indexFile.skip(10);
+
+	uint32 indexSize = indexFile.size();
+
+	if (filecnt > (indexSize - 12) / sizeof(FileDescriptorBin))
+		filecnt = (indexSize - 12) / sizeof(FileDescriptorBin);
+
+	for (uint i = 0; i < filecnt; i++) {
+		FileDescriptorBin descBin;
+		indexFile.read(&descBin, sizeof(descBin));
+		FileDescriptor desc(descBin);
+		files[desc._fileName] = desc;
+	}
+
+	return new MpsInstaller(files, baseName);
+}
+
+// Use of \\ as path separator is a guess. Never seen an archive with subfolders
+static Common::String translateName(const Common::Path &path) {
+	return Common::normalizePath(path.toString('\\'), '\\');
+}
+
+bool MpsInstaller::hasFile(const Common::Path &path) const {
+	return _files.contains(translateName(path));
+}
+
+int MpsInstaller::listMembers(Common::ArchiveMemberList &list) const {
+	int members = 0;
+
+	for (Common::HashMap<Common::String, FileDescriptor, Common::IgnoreCase_Hash, Common::IgnoreCase_EqualTo>::const_iterator i = _files.begin(), end = _files.end();
+	     i != end; ++i) {
+		list.push_back(Common::ArchiveMemberList::value_type(new Common::GenericArchiveMember(i->_key, this)));
+		++members;
+	}
+
+	return members;
+}
+
+const Common::ArchiveMemberPtr MpsInstaller::getMember(const Common::Path &path) const {
+	Common::String translated = translateName(path);
+	if (!_files.contains(translated))
+		return nullptr;
+
+	return Common::SharedPtr<Common::ArchiveMember>(new Common::GenericArchiveMember(_files.getVal(translated)._fileName, this));
```

Comment:
```suggestion
	return Common::SharedPtr<Common::ArchiveMember>(new Common::GenericArchiveMember(_files.getVal(translated)._fileName, this));
```
    return Common::ArchiveMemberPtr(new Common::GenericArchiveMember(name, this));

Refer to https://github.com/scummvm/scummvm/blob/253a0519a7d403980d265739741155f3f27ba00d/common/installshieldv3_archive.cpp#L137